tests/std/tests/P2231R1_complete_constexpr_optional_variant/test.cpp

```diff
@@ -0,0 +1,323 @@
+
+        variant<Dummy, T> move_assigned;
+        assert(move_assigned.index() == 0);
+        move_assigned = move(constructed);
+        assert(move_assigned.index() == 1);
+        assert(get<T>(move_assigned) == 42);
+    }
+
+    { // emplace type
+        T input{42};
+        variant<Dummy, T> emplace_copy;
+        assert(emplace_copy.index() == 0);
+        emplace_copy.template emplace<T>(input);
+        assert(emplace_copy.index() == 1);
+        assert(get<T>(emplace_copy) == 42);
+
+        variant<Dummy, T> emplace_move;
+        assert(emplace_move.index() == 0);
+        emplace_move.template emplace<T>(T{42});
+        assert(emplace_move.index() == 1);
+        assert(get<T>(emplace_move) == 42);
+
+        variant<Dummy, T> emplace_conversion;
+        assert(emplace_conversion.index() == 0);
+        emplace_conversion.template emplace<T>(42);
+        assert(emplace_conversion.index() == 1);
+        assert(get<T>(emplace_conversion) == 42);
+
+        variant<Dummy, T> emplace_initializer_list;
+        assert(emplace_initializer_list.index() == 0);
+        emplace_initializer_list.template emplace<T>({42, 43});
+        assert(emplace_initializer_list.index() == 1);
+        assert(get<T>(emplace_initializer_list) == 42);
+    }
+
+    { // emplace index
+        T input{42};
+        variant<Dummy, T> emplace_copy;
+        assert(emplace_copy.index() == 0);
+        emplace_copy.template emplace<1>(input);
+        assert(emplace_copy.index() == 1);
+        assert(get<T>(emplace_copy) == 42);
+
+        variant<Dummy, T> emplace_move;
+        assert(emplace_move.index() == 0);
+        emplace_move.template emplace<1>(T{42});
+        assert(emplace_move.index() == 1);
+        assert(get<T>(emplace_move) == 42);
+
+        variant<Dummy, T> emplace_conversion;
+        assert(emplace_conversion.index() == 0);
+        emplace_conversion.template emplace<1>(42);
+        assert(emplace_conversion.index() == 1);
+        assert(get<T>(emplace_conversion) == 42);
+
+        variant<Dummy, T> emplace_initializer_list;
+        assert(emplace_initializer_list.index() == 0);
+        emplace_initializer_list.template emplace<1>({42, 43});
+        assert(emplace_initializer_list.index() == 1);
+        assert(get<T>(emplace_initializer_list) == 42);
+    }
+
+    { // swap
+        variant<Dummy, T> left{T{42}};
+        variant<Dummy, T> right;
+        assert(left.index() == 1);
+        assert(get<T>(left) == 42);
+        assert(right.index() == 0);
+
+        left.swap(right);
+        assert(right.index() == 1);
+        assert(get<T>(right) == 42);
+        assert(left.index() == 0);
+
+        swap(left, right);
+        assert(left.index() == 1);
+        assert(get<T>(left) == 42);
+        assert(right.index() == 0);
+    }
+    return true;
+}
+
+int main() {
+    test_optional<With_trivial_destrutor>();
+    test_optional<With_nontrivial_destrutor>();
+#if defined(__cpp_lib_constexpr_dynamic_alloc) \

```

Comment:
I believe we can drop the `__cpp_lib_constexpr_dynamic_alloc`

This also explains why the tests are passing. They are not active for MSVC